### PR TITLE
Airspeed: Sanity check MS4525DO data, and utilize a double read

### PR DESCRIPTION
**Pull request by WickedShell**
It is possible to get a really incorrect data read off the I2C bus, that will come out with either wildly different numbers or a maximum value. Given that this is an incredibly illogical input for airspeed data, we are better off rejecting the data.

To do this we double read the sensor data, the measurements will drift slightly between two readings, but not enough that the upper byte will undergo a change. We also check that no reading was the maximum or minimum value possible.

This needs to be flight tested to ensure that we don't reject to many valid reasons, but given the required rate of change that should be fine.

The image below shows a rejected bad double read. Channel 0 is SDA, Channel 1 is SCL. Notice that the second read gets really bad data off the SCL line, which results in an erroneous read on the Pixhawk. The double pulse on channel 2 indicates that the check on line 143 failed. In this case our second reading was bad, but it can easily be the other way around as well.

rejected-bad-double-read

### CODE REVIEW

**Comment by tridge** [comment]
looks good to me, thanks michael!
just needs a flight test

File: libraries/AP_Airspeed/AP_Airspeed_MS4525.cpp [hunk 1]
```diff
@@ -120,6 +126,24 @@ void AP_Airspeed_MS4525::_collect()
     dT_raw = (data[2] << 8) + data[3];
     dT_raw = (0xFFE0 & dT_raw) >> 5;
 
+    int16_t dp_raw2, dT_raw2;
+    dp_raw2 = (data2[0] << 8) + data2[1];
+    dp_raw2 = 0x3FFF & dp_raw2;
+    dT_raw2 = (data2[2] << 8) + data2[3];
+    dT_raw2 = (0xFFE0 & dT_raw2) >> 5;
+
+    // reject any values that are the absolute minum or maximums
```
Comment (OXINARF) [inline libraries/AP_Airspeed/AP_Airspeed_MS4525.cpp]: minum -> minimum

File: libraries/AP_Airspeed/AP_Airspeed_MS4525.cpp [hunk 2]
```diff
@@ -120,6 +126,24 @@ void AP_Airspeed_MS4525::_collect()
...
+    // these can happen due to gnd lifts or communication erros on the bus
```
Comment (OXINARF) [inline libraries/AP_Airspeed/AP_Airspeed_MS4525.cpp]: erros -> errors

File: libraries/AP_Airspeed/AP_Airspeed_MS4525.cpp [hunk 3]
```diff
@@ -120,6 +126,24 @@ void AP_Airspeed_MS4525::_collect()
...
+    if (dp_raw  == 0x3FFF || dp_raw  == 0 || dT_raw  == 0x7FF || dT_raw == 0 ||
+        dp_raw2 == 0x3FFF || dp_raw2 == 0 || dT_raw2 == 0x7FF || dT_raw2 == 0) {
+        return;
+    }
+
+    // reject any double reads where the value has shifted in the upper more then a byte as corrupted
```
Comment (OXINARF) [inline libraries/AP_Airspeed/AP_Airspeed_MS4525.cpp]: The sentence looks like a mix of two. I would change it to:

reject any double reads where the values have a difference bigger than 255 (one lower byte)

File: libraries/AP_Airspeed/AP_Airspeed_MS4525.cpp [hunk 4]
```diff
@@ -120,6 +126,24 @@ void AP_Airspeed_MS4525::_collect()
...
+    if (abs(dp_raw - dp_raw2) >= 0xFF || abs(dT_raw - dT_raw2) > 0xFF) {
```
Comment (OXINARF) [inline libraries/AP_Airspeed/AP_Airspeed_MS4525.cpp]: Any reason why one uses >= and the other >? In my opinion should be only >

**Comment by WickedShell** [comment]
@OXINARF you might convince me to find a spell checking plugin for vim... :)

**Comment by WickedShell** [comment]
Flight tested to 20 m/s, worked quite well no periods of data outages, or visible anomalous readings.

**Comment by tridge** [comment]
applied, thanks Michael!

**Pull request merged**